clang-tools-extra/clang-tidy/readability/UseBuiltinLiteralsCheck.cpp

```diff
@@ -0,0 +1,171 @@
+  bool operator()(const LangOptions &LO) const {
+    return Std ? Std(LangStandard::getLangStandardForKind(LO.LangStd)) : true;
+  }
+  StringRef Seq;
+  RuleOnStd Std;
+};
+
+llvm::Regex CharRegex("^(u8|u|U|L)?");
+llvm::StringMap<Replacement> CharPrefix({
+    {"char", {""}},
+    {"char8_t", {"u8"}},
+    {"char16_t", {"u"}},
+    {"char32_t", {"U"}},
+    {"wchar_t", {"L"}},
+});
+
+llvm::Regex
+    IntRegex("(([uU]?[lL]{0,2})|([lL]{0,2}[uU]?)|([uU]?[zZ]?)|([zZ]?[uU]?))?$");
+llvm::StringMap<Replacement> IntSuffix({
+    {"int", {""}},
+    {"unsigned int", {"u"}},
+    {"long", {"L"}},
+    {"unsigned long", {"uL"}},
+    {"long long", {"LL"}},
+    {"unsigned long long", {"uLL"}},
+    {"size_t", {"uz", [](const auto &LS) { return LS.isCPlusPlus23(); }}},
+    {"std::size_t", {"uz", [](const auto &LS) { return LS.isCPlusPlus23(); }}},
+});
+
+llvm::Regex FloatRegex(
+    "([fF]|[lL]|([fF]16)|([fF]32)|([fF]64)|([fF]128)|((bf|BF)16))?$");
+llvm::StringMap<Replacement> FloatSuffix({
+    {"double", {""}},
+    {"float", {"f"}},
+    {"long double", {"L"}},
+    {"std::float16_t", {"f16"}},
+    {"std::float32_t", {"f32"}},
+    {"std::float64_t", {"f64"}},
+    {"std::float128_t", {"f128"}},
+    {"std::bfloat16_t", {"bf16"}},
+    {"float16_t", {"f16"}},
+    {"float32_t", {"f32"}},
+    {"float64_t", {"f64"}},
+    {"float128_t", {"f128"}},
+    {"bfloat16_t", {"bf16"}},
+});
+
+} // namespace
+
+void UseBuiltinLiteralsCheck::registerMatchers(MatchFinder *Finder) {
+  Finder->addMatcher(
+      traverse(TK_IgnoreUnlessSpelledInSource,
+               explicitCastExpr((has(ignoringParenImpCasts(
+                                    anyOf(characterLiteral().bind("char"),
+                                          integerLiteral().bind("int"),
+                                          floatLiteral().bind("float"))))))
+                   .bind("expr")),
+      this);
+  Finder->addMatcher(
+      traverse(TK_IgnoreUnlessSpelledInSource,
+               explicitCastExpr(has(initListExpr(has(ignoringParenImpCasts(
+                                    anyOf(characterLiteral().bind("char"),
+                                          integerLiteral().bind("int"),
+                                          floatLiteral().bind("float")))))))
+                   .bind("expr")),
+      this);
+}
+
+void UseBuiltinLiteralsCheck::check(const MatchFinder::MatchResult &Result) {
+
+  const auto &Sources = *Result.SourceManager;
+  const auto &Nodes = Result.Nodes;
+
+  const auto *MatchedCast = Nodes.getNodeAs<ExplicitCastExpr>("expr");
+  assert(MatchedCast);
+
+  std::string CastType = MatchedCast->getTypeAsWritten().getAsString();
+  std::string Fix; // Replacement string for the fix-it hint.
+  StringRef Seq;   // Literal sequence, prefix or suffix.
+
+  if (const auto *CharLit = Nodes.getNodeAs<CharacterLiteral>("char");
+      CharLit && CharPrefix.contains(CastType)) {
+    if (const auto &Rep = CharPrefix.at(CastType); Rep(getLangOpts())) {
```

Comment:
`auto` is only allowed to be used in two instances. If the type of the variable is way too complex to be spelt (e.g. some sort of a `std::container<foo, bar, baz, std::allocator<baz>, comparator, whatever, small_optimizer>::iterator_trait_impl<true>`) or if it is **trivial** from the initialisation expression. This is neither in this case.

```suggestion
    if (const Replacement &Rep = CharPrefix.at(CastType); Rep(getLangOpts())) {
```